src/net.c

```diff
@@ -2939,37 +2945,70 @@ void coap_cleanup(void) {
 
 #if ! defined WITH_CONTIKI && ! defined WITH_LWIP && ! defined RIOT_VERSION
 int
-coap_join_mcast_group(coap_context_t *ctx, const char *group_name) {
-  struct ipv6_mreq mreq;
+coap_join_mcast_group_intf(coap_context_t *ctx, const char *group_name,
+  const char *ifname) {
+  struct ip_mreq mreq4;
+  struct ipv6_mreq mreq6;
   struct addrinfo   *reslocal = NULL, *resmulti = NULL, hints, *ainfo;
+  struct ifreq ifr;
   int result = -1;
   coap_endpoint_t *endpoint;
   int mgroup_setup = 0;
 
-  /* we have to resolve the link-local interface to get the interface id */
-  memset(&hints, 0, sizeof(hints));
-  hints.ai_family = AF_INET6;
-  hints.ai_socktype = SOCK_DGRAM;
+  mreq6.ipv6mr_interface = -1U;
+  mreq4.imr_interface.s_addr = INADDR_BROADCAST;
 
-  result = getaddrinfo("::", NULL, &hints, &reslocal);
-  if (result != 0) {
-    coap_log(LOG_ERR,
-             "coap_join_mcast_group: cannot resolve link-local interface: %s\n",
-             gai_strerror(result));
-    goto finish;
-  }
+  if (!ifname) {
+    /* no interface specified */
+    /* we have to resolve the link-local interface to get the interface id */
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_INET6;
+    hints.ai_socktype = SOCK_DGRAM;
 
-  /* get the first suitable interface identifier */
-  for (ainfo = reslocal; ainfo != NULL; ainfo = ainfo->ai_next) {
-    if (ainfo->ai_family == AF_INET6) {
-      mreq.ipv6mr_interface =
+    result = getaddrinfo("::", NULL, &hints, &reslocal);
+    if (result != 0) {
+      coap_log(LOG_WARNING, "coap_join_mcast_group_intf: "
+               "cannot resolve link-local interface: %s\n",
+               gai_strerror(result));
+    }
+    else {
+      /* get the first suitable interface identifier */
+      for (ainfo = reslocal; ainfo != NULL; ainfo = ainfo->ai_next) {
+        if (ainfo->ai_family == AF_INET6) {
+            mreq6.ipv6mr_interface =
                 ((struct sockaddr_in6 *)ainfo->ai_addr)->sin6_scope_id;
-      break;
+          break;
+        }
+      }
+    }
+
+    mreq4.imr_interface.s_addr = INADDR_ANY;
+  }
+  else {
+    /* interface specified */
+    /* get interface index for IPv6 and address for IPv4 */
+    strncpy(ifr.ifr_name, ifname, IFNAMSIZ - 1);
+
+    result = ioctl(ctx->endpoint->sock.fd, SIOCGIFINDEX, &ifr);
+    if (result != 0) {
+      coap_log(LOG_WARNING, "coap_join_mcast_group_intf: "
+               "cannot get interface index for '%s': %d\n", ifname, result);
+    }
+    else {
+      mreq6.ipv6mr_interface = ifr.ifr_ifindex;
+    }
+
+    result = ioctl(ctx->endpoint->sock.fd, SIOCGIFADDR, &ifr);
```

Comment:
`SIOCGIFADDR` is for `AF_INET` sockets only, therefore it will always fail.

Reply:
What if `coap_new_endpoint` is passed a `listen_addr` that contains an `AF_INET` address? Then the socket will also be a `AF_INET` socket.